### ISSUE TITLE
Require the Nix daemon and client to be in sync

### ISSUE THREAD

**Issue by edolstra** [NixOS/nix#5685]
The requirement to be backward/forward-compatible in the Nix daemon/client adds a lot of cruft to Nix, and a lot of testing complexity. We can get rid of all of that if we require the daemon and client to be from the same version.

The only reason historically for supporting older/newer clients is `nixos-rebuild`: we want to evaluate the NixOS config using the latest Nix version because Nixpkgs may require it, and we *do* want to use the daemon to prevent a database schema upgrade before we actually switch to the new configuration. However, by policy Nixpkgs doesn't require the latest Nix version, so it should be okay to use the Nix version from the current system configuration.

Requiring the daemon and client to be the same version would be annoying for ssh-ng:// users. But most people use ssh:// which uses a much simpler protocol.

**Comment by edolstra** [comment]
CC @Ericson2314, @regnat.

**Comment by cole-h** [comment]
As nice as it would be to ditch all the related cruft, I agree with @regnat. I can imagine future users flocking to the support channels with "I want to test out flakes, but it says something about a version mismatch?" and related queries.

Even drive-by testing a bugfix in Nix itself could become impossible unless you switch your entire system to use the daemon from whatever commit or PR exists.

Imagine an issue comes up in the future and having to bisect it across versions. Requiring lockstep versions would definitely become a pain point there.

**Comment by Ericson2314** [comment]
Yeah what we *can* do is not worry about breaking changes between releases, for we burned through a lot of protocol versions over the course of 2.4 development. But we absolutely should compatibility at least between adjacent releases, for the reasons @regnat said. If we go to 6 weeks releases, then we should maintain e.g. 1 year's worth of releases of protocol compatibility.

**Comment by Ericson2314** [comment]
I completely agree `ssh://` merely seems simpler because we are missing features. I think supporting both indefinitely is making them both quite buggy as real world usage is awkwardly split between them. I hope we get hydra to support `ssh-ng://` and deprecate `ssh://` soon. I think it would be good to get creative on how we might factor out helper functions in `RemoteStore` such that hydra wouldn't need to re-implement the marshaling logic completely.

**Comment by rickynils** [comment]
I'll chime in with the other commenters here. I was under the impression that `ssh://` is being deprecated in favor of `ssh-ng://`, and in that case, forcing server/client to be in sync also means that all remote builders have to be in sync with the clients. I think that is simply too restrictive.

I would love to see the following happen:

* Deprecate `ssh://` once Hydra supports `ssh-ng://`. Keep supporting `ssh://` for a while, with deprecation warnings.
* Continue supporting multiple client versions in `ssh-ng://`, but if it would make things easier, we could ditch support for clients beyond a certain age.
* Start working on formalising, documenting and maybe reworking the daemon/builder protocol, and in that process find a way to make the protocol versioning more maintainable than all the conditionals that are spread out in the code today?

**Comment by tomberek** [comment]
Closing as stale. Re-open if needed.
